parachain/pallets/ethereum-beacon-client/src/lib.rs

```diff
@@ -303,41 +336,51 @@ pub mod pallet {
 		}
 
 		#[pallet::call_index(2)]
-		#[pallet::weight(T::WeightInfo::import_execution_header())]
+		#[pallet::weight(T::WeightInfo::import_versioned_execution_header())]
 		#[transactional]
-		pub fn import_execution_header(
+		pub fn import_versioned_execution_header(
 			origin: OriginFor<T>,
-			update: HeaderUpdateOf<T>,
+			versioned_update: VersionedHeaderUpdate<T>,
 		) -> DispatchResult {
 			let _sender = ensure_signed(origin)?;
 
 			Self::check_bridge_blocked_state()?;
 
-			let slot = update.beacon_header.slot;
-			let block_hash = update.execution_header.block_hash;
+			let slot: u64;
+			let block_hash: H256;
+			let result: DispatchResult;
 
-			log::info!(
-				target: "ethereum-beacon-client",
-				"💫 Received header update for slot {}.",
-				slot
-			);
-
-			if let Err(err) = Self::process_header(update) {
-				log::error!(
-					target: "ethereum-beacon-client",
-					"💫 Header update failed with error {:?}",
-					err
-				);
-				return Err(err)
+			match versioned_update {
+				VersionedHeaderUpdate::Bellatrix(update) => {
+					slot = update.beacon_header.slot;
+					block_hash = update.execution_header.block_hash;
+					result = Self::process_header(update);
```

Comment:
There is a lot of duplicated code in `process_header` and `process_header_capella`. I think we can make better use of generics to avoid this.

Reply:
For example, I analysed the diff between those two functions, and they differ by only 1 statement:

```diff
-    let execution_root = merkleization::hash_tree_root_execution_header(execution_payload.clone())
-        .map_err(|_| Error::<T>::BlockBodyHashTreeRootFailed)?;
+    let execution_root =
+        merkleization::hash_tree_root_execution_header_capella(execution_payload.clone())
+            .map_err(|_| Error::<T>::BlockBodyHashTreeRootFailed)?;
+
```

Reply:
Yes, bad smell here. Thanks for point out and I'll do some refactoring accordingly.

Reply:
@vgeddes some refactoring in commit
https://github.com/Snowfork/snowbridge/pull/786/commits/fff98fc2c968ffac3b23936c6cb4bb672af898d3